### PR TITLE
Add Safe Ethers Adapter

### PR DESCRIPTION
**Pull request by rmeissner**
Adds a Ethers signer that proposes a Safe transaction to the interface when used

- [x] Add Ethers signer that pushes transactions to Safe service
- [x] Add example how to use signer
- [x] Clean up hardcoded constants
- [x] Add tests 
- [x] Update readme
- [x] Add method documentation

### CODE REVIEW

File: packages/safe-ethers-adapters/src/signer.ts [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    async estimateSafeTx(safe: string, safeTx: SafeTransactionDataPartial): Promise<BigNumber> {
+        const url = `${this.serviceUrl}/api/v1/safes/${safe}/multisig-transactions/estimations/`
+        try {
+            console.log({safeTx})
+            const resp = await this.network.post(url, safeTx)
+            return BigNumber.from(resp.data.safeTxGas)
+        } catch (e) {
+            console.log((e as AxiosError).response)
```
Comment (rmeissner) [inline packages/safe-ethers-adapters/src/signer.ts:30]: Remove

File: packages/safe-ethers-adapters/src/signer.ts [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+            contractTransactionHash: safeTxHash,
+            sender: signature.signer,
+            signature: signature.data
+        }
+        try {
+            console.log({data})
+            const resp = await this.network.post(url, data)
+            return resp.data
+        } catch (e) {
+            console.log((e as AxiosError).response)
```
Comment (rmeissner) [inline packages/safe-ethers-adapters/src/signer.ts:54]: I will remove this

File: packages/safe-ethers-adapters/src/signer.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    async proposeTx(safeAddress: string, safeTxHash: string, safeTx: SafeTransaction, signature: SafeSignature): Promise<String> {
+        const url = `${this.serviceUrl}/api/v1/safes/${safeAddress}/multisig-transactions/`
+        const data = {
+            ...safeTx.data,
+            contractTransactionHash: safeTxHash,
+            sender: signature.signer,
+            signature: signature.data
+        }
+        try {
+            console.log({data})
```
Comment (rmeissner) [inline packages/safe-ethers-adapters/src/signer.ts:50]: I will remove this

File: packages/safe-ethers-adapters/src/signer.ts [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+    constructor(serviceUrl: string, network?: AxiosInstance) {
+        this.serviceUrl = serviceUrl
+        this.network = network ?? axios;
+    }
+
+    async estimateSafeTx(safe: string, safeTx: SafeTransactionDataPartial): Promise<BigNumber> {
+        const url = `${this.serviceUrl}/api/v1/safes/${safe}/multisig-transactions/estimations/`
+        try {
+            console.log({safeTx})
```
Comment (rmeissner) [inline packages/safe-ethers-adapters/src/signer.ts:26]: I will remove this

File: packages/safe-ethers-adapters/src/signer.ts [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    safe: Safe
+
+    static async create(address: string, signer: Signer, service: SafeService, provider?: Provider): Promise<SafeEthersSigner> { 
+        const safe = await EthersSafe.create({ethers, safeAddress: address, providerOrSigner: signer})
+        return new SafeEthersSigner(safe, service, provider)
+    }
+    
+    constructor(safe: Safe, service: SafeService, provider?: Provider) {
+        super(safe.getAddress(), provider)
+        const createLibDeployment = getCreateCallDeployment()
```
Comment (rmeissner) [inline packages/safe-ethers-adapters/src/signer.ts:78]: Maybe I should add an explicit null check for the deployment here

File: packages/safe-ethers-adapters/src/signer.ts [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+            hash: safeTxHash,
+            from: this.address,
+            confirmations: 0,
+            wait: async (confirmations?: number): Promise<TransactionReceipt> => {
+                while (true) {
+                    try {
+                        const txDetails = await connectedService.getSafeTxDetails(safeTxHash)
+                        if (txDetails.transactionHash) {
+                            this._checkProvider("sendTransaction")
+                            const receipt = await this.provider!!.waitForTransaction(txDetails.transactionHash, confirmations)
```
Comment (rmeissner) [inline packages/safe-ethers-adapters/src/signer.ts:106]: Extract this logic into a method

File: packages/safe-ethers-adapters/src/signer.ts [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        const tx = await transaction
+        let operation = 0
+        let to = await tx.to
+        let data = (await tx.data)?.toString() ?? "0x"
+        let value = BigNumber.from(await tx.value ?? 0)
+        if (!to) {
+            to = this.createLibAddress
+            data = this.createLibInterface.encodeFunctionData("performCreate", [value, data])
+            value = BigNumber.from(0)
+            operation = 1
```
Comment (rmeissner) [inline packages/safe-ethers-adapters/src/signer.ts:137]: Use operation enum

File: packages/safe-ethers-adapters/src/utils.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import { SafeTransactionData } from "@gnosis.pm/safe-core-sdk/dist/src/utils/transactions/SafeTransaction"
+import { getCreateCallDeployment } from "@gnosis.pm/safe-deployments"
+import { providers, utils } from "ethers"
+
+export const createLibDeployment = getCreateCallDeployment()
+export const createLibAddress = createLibDeployment!!.defaultAddress
+export const createLibInterface = new utils.Interface(createLibDeployment!!.abi)
+
+const mapStatus = (receipt: providers.TransactionReceipt): number => {
+    const success = receipt.logs.find((log: any) => log.topics[0] === "0x442e715f626346e8c54381002da614f62bee8d27386535b2521ec8540898556e")
```
Comment (germartinez) [inline packages/safe-ethers-adapters/src/utils.ts:10]: Could you specify in a comment the operation that corresponds to `0x442e715f626346e8c54381002da614f62bee8d27386535b2521ec8540898556e`?

File: packages/safe-ethers-adapters/src/utils.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    return !!success ? 1 : 0
+}
+
+const mapContractAddress = (receipt: providers.TransactionReceipt, safeTx: SafeTransactionData): string => {
+    if (safeTx.to.toLowerCase() === createLibAddress.toLowerCase()) {
+        const creationLog = receipt.logs.find((log: any) => log.topics[0] === "0x4db17dd5e4732fb6da34a148104a592783ca119a1e7bb8829eba6cbadef0b511")
```
Comment (germartinez) [inline packages/safe-ethers-adapters/src/utils.ts:16]: Could you specify in a comment the operation that corresponds to `0x4db17dd5e4732fb6da34a148104a592783ca119a1e7bb8829eba6cbadef0b511`?

File: packages/safe-ethers-adapters/src/utils.ts (hunk 8 above)
Comment (germartinez) [reply packages/safe-ethers-adapters/src/utils.ts:10]: I guess it is `ExecutionSuccess(bytes32 txHash, uint256 payment)`

File: packages/safe-ethers-adapters/tests/signer.test.ts [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+        getAddress: sinon.fake.returns("some_safe_address"),
+        getChainId: sinon.fake.returns(1),
+      }
+      const service: any = { 
+        getSafeTxDetails : sinon.fake.returns(safeTxDetails)
+      }
+      const logs = [
+        {
+          topics: [
+            "0x23428b18acfb3ea64b08dc0c1d296ea9c09702c09083ca5272e64d115b687d23"
```
Comment (germartinez) [inline packages/safe-ethers-adapters/tests/signer.test.ts:244]: Same here, I guess related with a failed transaction

File: packages/safe-ethers-adapters/src/utils.ts (hunk 9 above)
Comment (germartinez) [reply packages/safe-ethers-adapters/src/utils.ts:16]: Maybe related with a sucessful deployment of a contract?

**Review by germartinez — APPROVED** [review]

**Pull request merged**